armi/nuclearDataIO/cccc/dif3d.py

```diff
@@ -0,0 +1,263 @@
+                self._data.threeD[param] = record.rwDouble(self._data.threeD[param])
+
+    def _rw4DRecord(self):
+        """
+        Record for the optimum overrelaxation factors. This record is only present if `NUMORP`
+        is greater than 0.
+
+        The parameters are stored as a dictionary under the attribute `fourD`. This
+        could be changed into a list in the future since this record represents groupwise
+        data.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        """
+        if self._data.twoD["NUMORP"] != 0:
+            omegaParams = [f"OMEGA{e}" for e in range(1, self.twoD["NUMORP"] + 1)]
+
+            with self.createRecord as record:
+                # Initialize the record if we're reading
+                if self._data.fourD is None:
+                    self._data.fourD = {omegaParam: None for omegaParam in omegaParams}
+
+                for omegaParam in omegaParams:
+                    self._data.fourD[omegaParam] = record.rwDouble(
+                        self._data.fourD[omegaParam]
+                    )
+
+    def _rw5DRecord(self):
+        """
+        Record for the axial coarse-mesh rebalance boundaries. Coarse mesh balancing is
+        disabled in DIF3D-VARIANT, so this record is only relevant for DIF3D-Nodal. This
+        record is only present if `NCMRZS` is greater than 0.
+
+        The parameters are stored as a dictionary under the attribute `fiveD`.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        """
+        if self._data.twoD["NCMRZS"] != 0:
+            zcmrcParams = [f"ZCMRC{e}" for e in range(1, self.twoD["NCMRZS"] + 1)]
+            nzintsParams = [f"NZINTS{e}" for e in range(1, self.twoD["NCMRZS"] + 1)]
+
+            with self.createRecord as record:
+                # Initialize the record if we're reading
+                if self._data.fiveD is None:
+                    self._data.fiveD = {zcmrcParam: None for zcmrcParam in zcmrcParams}
+                    self._data.fiveD = {
+                        nzintsParam: None for nzintsParam in nzintsParams
+                    }
+
+                for zcmrcParam in zcmrcParams:
+                    self._data.fourD[zcmrcParam] = record.rwDouble(
+                        self._data.fourD[zcmrcParam]
+                    )
+                for nzintsParam in nzintsParams:
+                    self._data.fourD[nzintsParam] = record.rwInt(
+                        self._data.fourD[nzintsParam]
+                    )
+
+    def readWrite(self):
+        msg = f"{'Reading' if 'r' in self._fileMode else 'Writing'} DIF3D binary data {self}"
+        runLog.info(msg)
+
+        self._rwFileID()
+        self._rw1DRecord()
+        self._rw2DRecord()
+        self._rw3DRecord()
+        self._rw4DRecord()
+        self._rw5DRecord()
+
+
+readBinary = Dif3dStream.readBinary  # pylint: disable=invalid-name
```

Comment:
Nice. It does look like many of the other readers include these convenience functions, but some of them do not; we might consider tracking that as a inconsistency to fix in the future.

Reply:
https://github.com/terrapower/armi/issues/1315